plugin/pkg/auth/authorizer/webhook/webhook.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package webhook implements the authorizer.Authorizer interface using HTTP webhooks.
+package webhook
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/authorization/v1beta1"
+	"k8s.io/kubernetes/pkg/auth/authorizer"
+	"k8s.io/kubernetes/pkg/util"
+)
+
+// Ensure Webhook implements the authorizer.Authorizer interface.
+var _ authorizer.Authorizer = (*WebhookAuthorizer)(nil)
+
+type WebhookAuthorizer struct {
+	callback string
+	cli      *http.Client
+}
+
+func New(callback, clientCertFile, clientKeyFile, caFile string) (*WebhookAuthorizer, error) {
+	var clientTLSConfig tls.Config
+
+	if clientCertFile != "" && clientKeyFile != "" {
+		clientCert, err := tls.LoadX509KeyPair(clientCertFile, clientKeyFile)
+		if err != nil {
+			return nil, err
+		}
+		clientTLSConfig.Certificates = []tls.Certificate{clientCert}
+	} else {
+		glog.Info("No x509 certificate provided for client identity.")
+	}
+
+	if caFile != "" {
+		roots, err := util.CertPoolFromFile(caFile)
+		if err != nil {
+			return nil, err
+		}
+		clientTLSConfig.RootCAs = roots
+	} else {
+		glog.Info("No x509 certificates provided, will use host's root CA set")
+	}
+
+	// TODO(ericchiang): Can we check here to ensure remote service is reachable?
+
+	return &WebhookAuthorizer{
+		callback: callback,
+		cli: &http.Client{
+			// Same as http.DefaultTransport except for TLSClientConfig.
+			Transport: &http.Transport{
+				Proxy: http.ProxyFromEnvironment,
+				Dial: (&net.Dialer{
+					Timeout:   30 * time.Second,
+					KeepAlive: 30 * time.Second,
+				}).Dial,
+				TLSHandshakeTimeout: 10 * time.Second,
+				TLSClientConfig:     &clientTLSConfig,
+			},
+		},
+	}, nil
+}
+
+// RequestError indicates an error composing, sending, or decoding the response of a request.
+type RequestError struct {
+	msg string
+}
+
+func (r *RequestError) Error() string {
+	return r.msg
+}
+
+func logErrorf(format string, a ...interface{}) *RequestError {
+	msg := fmt.Sprintf(format, a...)
+	glog.Error(msg)
+	return &RequestError{msg}
+}
+
+func (w *WebhookAuthorizer) Authorize(attr authorizer.Attributes) (err error) {
+	r := v1beta1.SubjectAccessReviewSpec{
+		User:   attr.GetUserName(),
+		Groups: attr.GetGroups(),
```

Comment:
If you are saying you always know the groups from Authentication step, then I am okay if you leave this code alone.  We will need this feature but I guess we can sort it out when we are ready.